### PR TITLE
PBFT - share the underlying Message buffer

### PR DESCRIPTION
**Pull request by ashamis**

### CODE REVIEW

**Comment by cimetrics** [comment]
shared_buffer_5@[6067 aka 20200318.4](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6067&view=results) vs master ewma over 30 builds from [5772](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5772&view=results) to [6062](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6062&view=results)
images

File: src/consensus/pbft/libbyz/Message.cpp [hunk 1]
```diff
@@ -15,23 +15,25 @@ std::unique_ptr<Log_allocator> thread_allocator = nullptr;
...
 {
   if (sz != 0)
   {
     if (thread_allocator == nullptr)
     {
       thread_allocator = std::make_unique<Log_allocator>();
     }
-    allocator = thread_allocator.get();
+    auto allocator = thread_allocator.get();
+    int max_size = (ALIGNED_SIZE(sz));
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/Message.cpp]: size_t, no extra brackets?

File: src/consensus/pbft/libbyz/Message.cpp [hunk 2]
```diff
@@ -15,23 +15,25 @@ std::unique_ptr<Log_allocator> thread_allocator = nullptr;
...
 
-    msg = (Message_rep*)allocator->malloc(max_size);
-    if (msg != nullptr)
+    auto m = (Message_rep*)allocator->malloc(max_size);
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/Message.cpp]: C cast

File: src/consensus/pbft/libbyz/Message.h [hunk 3]
```diff
@@ -119,6 +119,34 @@ class Message
...
+    {
+      if (max_size > 0 && msg != nullptr)
+      {
+        allocator->free((char*)msg, max_size);
+      }
+    }
+
+    Message_rep* msg; // Pointer to the contents of the message.
+    Log_allocator* allocator;
+    int max_size; // Maximum number of bytes that can be stored in "msg"
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Message.h]: `size_t`

File: src/consensus/pbft/libbyz/Message.h [hunk 4]
```diff
@@ -119,6 +119,34 @@ class Message
...
+                  // or "-1" if this instance is not responsible for
+                  // deallocating the storage in msg.
+  };
+
+  std::shared_ptr<MsgBufCounter> get_msg_buffer()
+  {
+    return msg;
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Message.h]: Should the `msg` member be `private:`?

File: src/consensus/pbft/libbyz/Message.h [hunk 5]
```diff
@@ -167,20 +190,22 @@ class Message
 
 inline int Message::size() const
 {
-  return msg->size;
+  return msg->msg->size;
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Message.h]: `msg->get_msg_buffer()`?

**Comment by ashamis** [comment]
I am going to hold off on this PR for a little bit. Something is off perf wise and I need to understand what it is.

**Pull request closed without merging**